File: 14-parabolic-reflector-dish/main.py
from enum import Enum
from typing import List, Set, Tuple
from dataclasses import dataclass, replace
from itertools import count, repeat
from functools import reduce
from copy import copy
from collections.abc import Callable
import sys


# magic
def tailrec(g: Callable) -> Callable:

    class TailRecurseException(BaseException):
        def __init__(self, args, kwargs):
            self.args = args
            self.kwargs = kwargs

    def func(*args, **kwargs):
        f = sys._getframe()
        if f.f_back and f.f_back.f_back and f.f_back.f_back.f_code == f.f_code:
            raise TailRecurseException(args, kwargs)
        else:
            while True:
                try: return g(*args, **kwargs)
                except TailRecurseException as e:
                    args = e.args
                    kwargs = e.kwargs

    func.__doc__ = g.__doc__
    return func


class Direction(Enum):
    NORTH = 1
    SOUTH = 2
    EAST = 3
    WEST = 4


Coords = Tuple[int, int]


@dataclass
class Dish:
    rounded: Set[Coords]  # 'O'
    cubed: Set[Coords]    # '#'
    height: int
    width: int

    def has_obstacle_at(self, coords: Coords) -> bool:
        return (coords in self.rounded) or (coords in self.cubed)


# 1, 'O.OO#....#' -> O: (1, 0), (1, 2), (1, 3)...
def parse_line(n: int, line: str) -> Tuple[Set[Coords], Set[Coords]]:
    o_indices = [e for e, c in enumerate(line) if c == 'O']
    hash_indices = [e for e, c in enumerate(line) if c == '#']
    o_coords = zip(repeat(n), o_indices)
    hash_coords = zip(repeat(n), hash_indices)
    return (set(o_coords), set(hash_coords))


def parse_dish(txt: str) -> Dish:
    lines = txt.splitlines()
    [head, *_] = lines
    width = len(head)
    height = len(lines)
    pairs = map(parse_line, count(), lines)
    round: Tuple[Set[Coords]]
    cube: Tuple[Set[Coords]]
    round, cube = zip(*pairs)
    empty: Set[Coords] = set()
    round_coords = reduce(lambda acc, x: acc.union(x), round, empty)
    cube_coords = reduce(lambda acc, x: acc.union(x), cube, empty)
    return Dish(width=width, height=height, rounded=round_coords, cubed=cube_coords)


def next_coords(dir: Direction, coords: Coords) -> Coords:
    (row, col) = coords
    match dir:
        case Direction.NORTH: return (row - 1, col)
        case Direction.SOUTH: return (row + 1, col)
        case Direction.EAST: return (row, col + 1)
        case Direction.WEST: return (row, col - 1)


def out_of_bounds(dish: Dish, coords: Coords) -> bool:
    (row, col) = coords
    return row < 0 or col < 0 or row >= dish.height or col >= dish.width


@tailrec
def move(dir: Direction, dish: Dish, coords: Coords) -> Coords:
    new = next_coords(dir, coords)
    if dish.has_obstacle_at(new) or out_of_bounds(dish, new):
        return coords
    else:
        return move(dir, dish, new)


def after_move(dir: Direction, dish: Dish, coords: Coords) -> Dish:
    replica = copy(dish.rounded)
    replica.remove(coords)
    new = move(dir, dish, coords)
    replica.add(new)
    return replace(dish, rounded=replica)


def move_order(dir: Direction, coordss: Set[Coords]) -> List[Coords]:
    l = list(coordss)
    match dir:
        case Direction.NORTH: return sorted(l)
        case Direction.SOUTH: return sorted(l, reverse=True)
        case Direction.WEST: return sorted(l, key=lambda t: t[1])
        case Direction.EAST: return sorted(l, key=lambda t: t[1], reverse=True)


def tilt(dir: Direction, dish: Dish) -> Dish:
    order = move_order(dir, dish.rounded)

    @tailrec
    def aux(dir: Direction, dish: Dish, ord: List[Coords]) -> Dish:
        match ord:
            case [c, *cs]:
                new_dish = after_move(dir, dish, c)
                return aux(dir, new_dish, cs)
            case _:
                return dish

    return aux(dir, dish, order)


def load(dish: Dish) -> int:
    
    def load1(c: Coords) -> int:
        (row, _) = c
        return dish.height - row

    return sum(map(load1, dish.rounded))


def tilt_cycle(dish: Dish) -> Dish:
    return tilt(Direction.EAST, tilt(Direction.SOUTH, tilt(Direction.WEST, tilt(Direction.NORTH, dish))))


def main_():
    handle = open('input.txt', 'r')
    input = handle.read()
    handle.close()
    dish = parse_dish(input)
    tilted = tilt(Direction.NORTH, dish)
    total = load(tilted)
    print(total)


def main():
    handle = open('input.txt', 'r')
    input = handle.read()
    handle.close()
    dish = parse_dish(input)

    out = open('out.txt', 'w')
    out.write('')
    out.close()

    out = open('out.txt', 'a')
    for i in range(1000):
        out.write(str(i) + ' ' + str(load(dish)))
        out.write('\n')
        out.flush()
        dish = tilt_cycle(dish)

    # now find the pattern


if __name__ == '__main__':
    main()

File: 14-parabolic-reflector-dish/app/Main.hs
{-# LANGUAGE RecordWildCards #-}

module Main (main', main) where

import           Data.Bifunctor (Bifunctor (first, second))
import           Data.List      (elemIndices, sort, sortOn)
import           Data.Ord       (Down (Down))
import           Data.Set       (Set, delete, fromList, insert, member,
                                 toAscList, toDescList)
import           Data.Tuple     (swap)
import           System.IO      (IOMode (WriteMode), hClose, hPrint, openFile)

data Direction = North | South | East | West
  deriving (Eq, Show)

type Coords = (Int, Int)

data Dish = Dish { rounded :: Set Coords
                 , cubed   :: Set Coords
                 , height  :: Int
                 , width   :: Int
                 } deriving (Show, Eq)

instance Semigroup Dish where
  a <> b = Dish { rounded = r, cubed = c, height = h, width = w }
    where r = rounded a <> rounded b
          c = cubed a <> cubed b
          h = height a + height b
          w = max (width a) (width b)

instance Monoid Dish where
  mempty = Dish mempty mempty 0 0

-- Tilt the platform so that the round rocks all roll <North>
-- BUGGED:
tilt :: Direction -> Dish -> Dish
tilt dir dish = tilt' (toList $ rounded dish) dir dish
  where toList = case dir of
                   North -> toAscList
                   South -> toDescList
                   West  -> map swap . sort . map swap . toAscList
                   East  -> map swap . sortOn Down . map swap . toAscList

tilt' :: [Coords]  -- what's left to move?
      -> Direction -> Dish -> Dish
tilt' [] _ d = d
tilt' (c:cs) dir dish = tilt' cs dir after
  where after = afterMove dir dish c

afterMove :: Direction -> Dish -> Coords -> Dish
afterMove dir d old = d{rounded=r}
  where r = insert new $ delete old (rounded d)
        new = move dir d old

-- given all the context,
-- move a rock at given coords to the furthest
-- on a direction
-- until it reaches the edge
-- or run into another rock.
move :: Direction -> Dish -> Coords -> Coords
move dir dish old
  | newCoords `isOutOfBoundOf` dish = old
  | dish `hasRockAt` newCoords = old
  | otherwise = move dir dish newCoords
  where newCoords = nextCoords dir old

nextCoords :: Direction -> Coords -> Coords
nextCoords North = first $ subtract 1
nextCoords South = first (+1)
nextCoords West  = second $ subtract 1
nextCoords East  = second (+1)

isOutOfBoundOf :: Coords -> Dish -> Bool
isOutOfBoundOf (row, col) Dish{..} = row < 0 || row >= height || col < 0 || col >= width

hasRockAt :: Dish -> Coords -> Bool
d `hasRockAt` c = (d `byCoords` c) /= '.'

byCoords :: Dish -> Coords -> Char
byCoords Dish{..} c
  | c `member` rounded = 'O'
  | c `member` cubed = '#'
  | otherwise = '.'

-- total load on the <North> support beams
load :: Direction -> Dish -> Int
load North Dish{..} = foldr ((+) . load1) 0 rounded
  where load1 (row, _) = height - row
load _     _ = error "todo"

parseDish :: String -> Dish
parseDish = do lined <- lines
               let enum = zip [0..] lined
               let dishes = map (uncurry parseLine) enum
               return $ mconcat dishes

parseLine :: Int     -- curr line number
          -> String
          -> Dish
parseLine k str = Dish os pounds 1 (length str)
  where os = coords 'O'
        pounds = coords '#'
        coords c = fromList $ map (k,) $ elemIndices c str

main' :: IO ()
main' = do input <- readFile "test.txt"
           let dish = parseDish input
           let tilted = tilt North dish
           print $ load North tilted

-- part 2

tiltCycle :: Dish -> Dish
tiltCycle = tilt East . tilt South . tilt West . tilt North

-- try find the pattern yourself :)
main :: IO ()
main = do input <- readFile "input.txt"
          output <- openFile "output.txt" WriteMode
          let dish = parseDish input
          let tilteds = iterate tiltCycle dish
          let first1000 = map (load North) (take 1000 tilteds)
          let enum100 = zip [(0 :: Int)..] first1000
          mapM_ (hPrint output) enum100
          hClose output
